docs/animation/asm.md

```diff
@@ -0,0 +1,312 @@
+
+Each parameter can either be defined in the ASM or at load-time. Load-time parameters are usually of the type `VariableValue`, which returns a value changeable in-code, ignoring its input.
+Other kinds are `SimpleExprValue`, allowing you to do calculations on its input, `IdentityValue`, which just returns its input, `ConstValue` which returns a constant, `CompositionValue`, which chains parameters together and `ParameterValue`, which returns whatever anoter parameter _would_ return.
+A parameter is an input to a clip, telling it how far into the animation to animate.
+
+
+### Clips
+
+!!! note
+    
+    Clips can either be ASM-clips, ones that are defined in the ASM, or armature-clips, ones that are defined in the armature file.
+    For the rest of this page, "clips" will refer to ASM-clips unless otherwise stated
+
+A clip takes in an input, usually the time and does something to the model with it. Differnet types of clips do different things with the input, the easiest to use being the `TimeClip` and `ModelClip`. A `TimeClip` animates a given clip by a given parameter, usually that clip is a
+`ModelClip`, a reference to an armature-clip defined in an armature file.
+Other kinds include the `IdentityClip`, which does nothing, the `ClipReference`, which just does the same as another clip,
+the `SlerpClip`, which blends between two clips, useful for transitions, and the `TriggerClip`, used to both animate a clip and fire an event when its parameter goes positive.
+
+### Events
+
+Various things can trigger events in the ASM, but these are _not_ forge events. ASM events are simply strings. Events can be of two types, normal events and special events.
+Special events' text is formatted like this: `!event_type:event_value`. Right now there is only one kind of `event_type`, `transition`. This tries to transition to whatever state
+is defined in the `event_value`. Normal events can be _any other string_ and can be used from the `pastEvents` callback, but more information about that is on the implemting page, as
+that callback is in different places depending on what is being animated.
+
+
+Code API
+----------
+
+!!! warning
+
+    The ASM code API can only be used _client side_. When storing ASMs in code, use the side-irrelevant `IAnimationStateMachine` interface.
+   
+
+ASMs can be loaded by calling `ModelLoaderRegistry.loadASM`. It takes two parameters, first being a `ResourceLocation` denoting 
+where the ASM is stored, and second an ImmutableMap of load-time defined parameters.
+
+An example:
+```java
+@Nullable
+private final IAnimationStateMachine asm;
+private final VariableValue cycle = new VariableValue(4);
+
+public Spin() {
+    if (FMLCommonHandler.instance().getSide() == Side.CLIENT) {
+        asm = ModelLoaderRegistry.loadASM(new ResourceLocation(MODID, "asms/block/rotatest.json"), ImmutableMap.of("cycle_length", cycle));
```

Comment:
is this actually side safe? i'm not sure how loading of methods works but we don't want to show dediserver-crashing code in the official docs

Reply:
actually you are right, this should be proxy'd (oops)